### DISCUSSION TITLE
Workflow with UBL

### DISCUSSION THREAD

**Discussion by willie013** [Jyers/Marlin#308] [Q&A] [answered]
Hi everyone,

First of all, this Marlin project it great! It offers really nice features for the Ender 3 v2.
I am however a bit confused how the workflow should be with a BLTouch with UBL enabled.
Do I just have to create a UBL mesh on the printer using the level menu and then supply this Gcode in my slicer after G28?

G29 L1; load mesh
G29 J; probe 3 points to tilt mesh
M420 S1; restore ABL mesh

I am a little bit uncertain as to what command are executed when using the UBL menu in the LCD panel, and if these start GCODE commands are enough paired with this.
And after a few prints I can create a new mesh on the printer using the level menu and store it? Is this the correct workflow?

Thanks!

**Comment by Jyers** [comment, marked as answer]
That sounds about right to me. If you want more info you might want to check out Marlin's own documentation: https://marlinfw.org/docs/features/unified_bed_leveling.html It's quite a read, but has a lot of good information

**Reply by xantrk** [reply]
I believe that G29 A is also needed for convenience. After powercycle UBL gets disabled, and if it is not enabled before G29J explicitly, I get an invalid mesh afte  G29 J when I check via pronterface. After a couple tries, this is what works for me.

G28 ; home all
G29 A ; Activate UBL
G29 L ; Load the mesh stored, not sure if slot 1 is always correct
G29 J ; Probe 3 points to tilt mesh

BTW, I also saw comments on official marlin git that PROBING_MARGIN should be bigger than MESH_INSET, so that tilting probes are within the mesh itself.

**Reply by Jyers** [reply]
Thanks for the tip on probing margin, I'll definitely look at that

**Reply by danrough** [reply]
>BTW, I also saw comments on official marlin git that PROBING_MARGIN should be bigger than MESH_INSET, so that tilting probes are within the mesh itself.

I just came here to see if anyone else had reported an issue I am facing, which sounds very similar to this.

The `G29 J` sees the probe hit one of the stock metal bed clips. Is the short answer to my question of, how do I solve this - compile your own version of this firmware with the `PROBING_MARGIN` configured to be greater than the `MESH_INSET` value?

**Reply by Jyers** [reply]
I believe that would solve the issue. PROBING_MARGIN should set how far in the tilt grid is from the edges of the bed

**Reply by danrough** [reply]
Thanks!

**Reply by tommymonk** [reply]
I've been experiencing strange behaviour when using these two approaches;

Full probe
```
G28 ; Home all axes
G29 A; Activate the UBL system

G29 P1; Auto bed-level (BL-Touch)
G29 P3 T; Auto-fill unpopulated mesh
G29 S1; Save mesh to EEPROM Slot 1
```
Loading previous
```
G28 ; Home all axes
G29 A; Activate the UBL system

G29 L1; Load the mesh stored in slot 1
G29 J; No size specified on the J option tells G29 to probe the specified 3 points and tilt the mesh according to what it finds. 
```

Without any change to my z-offset or even touching the printer the three point levelling of the second code will be so low it scrapes my bed and the first version with full probe will print at the perfect height.

**Reply by Jyers** [reply]
The firmware stores the mesh in slot zero, not slot 1, so you need G29 L0

**Reply by tommymonk** [reply]
I didn't spot that, thanks @Jyers you're a star

**Comment by samsp-msft** [comment]
Can we get a definitive answer as to what needs to be put in the slicer's startup sequence to use the UBL, and then add it to the first page of the wiki?

**Reply by Jyers** [reply]
I've added some more information about UBL gcode to the configuration details wiki page, but to get a good grasp of UBL gcode as a whole I would recommend looking over this page https://marlinfw.org/docs/gcode/G029-ubl.html
